### ISSUE TITLE
Self Attestation is not supported

### ISSUE THREAD

**Issue by prusnak** [Yubico/libfido2#57]
It seems that [Self Attestation](https://www.w3.org/TR/webauthn/#self-attestation) is not supported, only Basic Attestation.

**Comment by prusnak** [comment]
Here's the culprit:

https://github.com/Yubico/libfido2/blob/5d0c77e5dc675762185ede1ef178286c7e2ab7d8/src/cred.c#L148

This code checks whether `x5c` field is provided and this is not filled in when using Self Attestation.

**Comment by martelletto** [comment]
Thank you for the feedback. Fixes for #57 and #58 should be ready in ~3 weeks; the chief reason being that I am on vacation until then. :)

-p.

**Comment by martelletto** [comment]
tentative support implemented in 5740caf8; will merge to master once the documentation bits are in place and the new code has been (at least) minimally fuzzed.

**Comment by martelletto** [comment]
(disclaimer: completely untested as of the moment)

**Comment by prusnak** [comment]
I tested the following (example from the fido2-cred manpage) with Trezor Model T (which does Self Attestation by default):

```
echo credential challenge | openssl sha256 -binary | base64 > cred_param
echo relying party >> cred_param
echo user name >> cred_param
dd if=/dev/urandom bs=1 count=32 | base64 >> cred_param
fido2-cred -M -i cred_param /dev/hidraw5 | fido2-cred -V -o cred
```

Master branch ends up with the following:

```
fido2-cred: fido_dev_make_cred: FIDO_ERR_INVALID_CBOR
fido2-cred: input error
```

(This is expected, that's why I filed the bug).

However 5740caf8ff2759d57d9d767462e62f21df96de6a ends up with the following:

```
fido2-cred: output error
fido2-cred: input error
```

Which is not expected.

Any ideas how can I help you to debug the problem?

**Comment by prusnak** [comment]
Now I realize the new function `fido_cred_verify_self` is not used anywhere yet. Plus the removal of check for `fido_blob_is_empty(&cred->attstmt.x5c)` in `fido_dev_make_cred_rx` with no replacement seems spurious.

**Comment by martelletto** [comment]
Yes; my apologies for the miscommunication. I have added a new option to fido2-cred (-s) in a8f47b778 to trigger fido_cred_verify_self() and enforce Self Attestation. (Untested as of the moment)

**Comment by prusnak** [comment]
I confirm that the following works correctly with Trezor Model T and a8f47b7 :

```
./fido2-cred -M -s -i cred_param /dev/hidraw3 | ./fido2-cred -V -s -o cred
```

**Comment by prusnak** [comment]
Quick question: Shouldn't the `fido2-cred` tool detect Self Attestation automatically rather than forcing a user to specify this using a command-line option?

**Comment by martelletto** [comment]
Thanks for testing!

I would also prefer if fido2-cred could handle Self Attestation transparently, but I don't see a reasonable way forward.

Technically, we could detect Self Attestation in both cases (-M and -V) by checking whether an attestation certificate is present. However, the attestation type used to verify the credential (-V) seems like an important enough aspect for the user to be explicitly aware of, hence the preference for a separate option. We could solve this by including the attestation type in the output produced by -V, but that would break scripts  relying in the current output.

If you have any insights, I would be glad to hear them!

-p.

**Comment by prusnak** [comment]
It would certainly be nice if scripts that use fido2-cred would still work even when devices with Self Attestation are used without the need to modify these scripts.

**Comment by martelletto** [comment]
Sure. I have implemented the necessary modifications, but unfortunately
can't push them as I don't have my work laptop with me.

I have placed a diff at https://ambientworks.net/tmp/fido2-cred.txt. If you
curl -LO then patch -p0 <, it should apply on self_attestation's HEAD.

Apologies for the inconvenience.

**Comment by prusnak** [comment]
I confirm that the following works with Trezor Model T (after applying the patch from above):

```
./fido2-cred -M -i cred_param /dev/hidraw3 | ./fido2-cred -V -o cred
```

(note that `-s` parameter is not used)

Great!

**Comment by martelletto** [comment]
Merged in 9522a3b; thanks for the testing and feedback!
